File: CPP Implementation/Quick_sort.cpp
#include<bits/stdc++.h>
using namespace std;

vector<int> quick_sort(vector<int> &arr)
{
  if(arr.size() <= 1) return arr;

  int pivot = rand() % arr.size();

  vector<int> a,b;

  for(int i = 0; i < arr.size(); i++)
  {
    if(i == pivot) continue;

    if(arr[i] <= arr[pivot])
    {
      a.push_back(arr[i]);
    }
    else
    {
      b.push_back(arr[i]);
    }
  }

  vector<int> sorted_a = quick_sort(a);
  vector<int> sorted_b = quick_sort(b);

  vector<int> sorted_arr;

  for(int i = 0; i < sorted_a.size(); i++)
  {
    sorted_arr.push_back(sorted_a[i]);
  }

  sorted_arr.push_back(arr[pivot]);

  for(int i = 0; i < sorted_b.size(); i++)
  {
    sorted_arr.push_back(sorted_b[i]);
  }

  return sorted_arr;
}

int main()
{
    vector<int> arr = {2,7,5,1,8,4};
    vector<int> sorted = quick_sort(arr);

    for(int i = 0; i < sorted.size(); i++)
    {
        cout<< sorted[i] << " ";
    }

    return 0;
}

File: Python Implementation/Quick_sort.py
from random import randint

def quick_sort(arr):
    n = len(arr)

    if n <= 1:
        return arr
    
    pivot = randint(0, n-1)

    a = []
    b = []

    for i in range(n):
        if i == pivot:
            continue

        if arr[i] <= arr[pivot]:
            a.append(arr[i])
        else:
            b.append(arr[i])
    
    sorted_a = quick_sort(a)
    sorted_b = quick_sort(b)

    sorted_arr = []

    for i in range(len(sorted_a)):
        sorted_arr.append(sorted_a[i])
    
    sorted_arr.append(arr[pivot])

    for i in range(len(sorted_b)):
        sorted_arr.append(sorted_b[i])

    return sorted_arr

arr = [2,7,5,1,8,4]

sorted = quick_sort(arr)

print(*sorted, sep=" ")